lib/core/src/Cardano/Wallet/Network.hs

```diff
@@ -377,21 +490,216 @@ instance ToText tr => ToText (FollowLog tr) where
+    -> UTCTime
+    -> FollowStats LogState
+    -> m (FollowStats LogState)
+flush tr t s = do
+    -- This is where we need to update the time
+    let s' = s { time = overCurrent (const t) (time s) }
+    traceWith tr s'
+    return $ hoistStats forgetPrev s'
+  where
+    forgetPrev (LogState _prev cur) = LogState cur cur
+
+emptyStats :: UTCTime -> SyncProgress -> FollowStats LogState
+emptyStats t p = FollowStats (f 0) (f 0) (f t) (f p)
+  where
+    f = initLogState
+
+-- | Update the stats based on a new log message.
+updateStats :: FollowLog tr -> FollowStats LogState -> FollowStats LogState
+updateStats msg s = force $ case msg of
+    MsgApplyBlocks _tip blocks ->
+        s { blocksApplied = overCurrent (+ NE.length blocks) (blocksApplied s) }
+    MsgDidRollback _ _ ->
+        s { rollbacks = overCurrent (1 +) (rollbacks s) }
+    MsgSyncProgress p ->
+        s { prog = overCurrent (const p) (prog s) }
+    _ -> s
+
+instance ToText (FollowStats LogState) where
+    toText st@(FollowStats b r t sp) = syncStatus <> " " <> stats <> sevExpl
+      where
+        -- NOTE: We are using NotResponding as initial value.
+        syncStatus = case sp of
+            LogState NotResponding Ready ->
+                "In sync."
+            LogState Ready Ready ->
+                "Still in sync."
+            LogState NotResponding NotResponding ->
+                "Still not syncing."
+            LogState (Syncing _p) Ready ->
+                "In sync!"
+            LogState Ready (Syncing p) ->
+                "Syncing again (" <> (pretty p) <> ")"
+            LogState _ (Syncing p) ->
+                "Syncing (" <> (pretty p) <> ")"
+            LogState prev NotResponding ->
+                "Not responding. Previously " <> (pretty prev) <> "."
+        stats = mconcat
+            [ "Applied " <> pretty (using (-) b) <> " blocks, "
+            , pretty (using (-) r) <> " rollbacks "
+            , "in the last " <> pretty (using diffUTCTime t) <> "."
+            ]
+          where
+            using f x = f (current x) (prev x)
+
+        sevExpl = maybe
+            ""
+            (\x -> " (" <> x <> ")")
+            (snd $ explainedSeverityAnnotation st)
+
+-- NOTE: Here we check if the sync progress is going backwards, which
+-- would be a sign the wallet is overloaded (or rollbacks)
+--
+-- But this check might be in the wrong place. Might be better to
+-- produce new logs from inside the updateStats function and immeditely
+-- warn there.
+explainedSeverityAnnotation :: FollowStats LogState -> (Severity, Maybe Text)
+explainedSeverityAnnotation s
+        | progressMovedBackwards = (Warning, Just "progress decreased")
+        | noBlocks && notRestored = (Warning, Just "not applying blocks")
+        | otherwise = (Info, Nothing)
+  where
+    progressMovedBackwards = current (prog s) < prev (prog s)
+    notRestored = current (prog s) /= Ready
+    noBlocks = (current (blocksApplied s) - prev (blocksApplied s)) <= 0
```

Comment:
I did this as a workaround for the problem described in the `prog :: f SyncProgress` doc  above.

But having the sync progress reported wrong (not re-calculated as the current time changes), could probably be really confusing. I should have a look at this again.

Reply:
This is now fixed by making it easy to calculate syncProgress without node-connection. Now it returns `NotResponding` rather than blocking.